### PR TITLE
rsa_pss_compute_saltlen(): Avoid integer overflows and check MD and RSA sizes

### PR DESCRIPTION
**Pull request by t8m**
Fixes Coverity 1604651

### CODE REVIEW

**Review by beldmit — APPROVED** [review]
LGTM

File: providers/implementations/signature/rsa_sig.c [hunk 1]
```diff
@@ -216,13 +216,29 @@ static int rsa_pss_compute_saltlen(PROV_RSA_CTX *ctx)
...
     }
     if (saltlen == RSA_PSS_SALTLEN_MAX || saltlen == RSA_PSS_SALTLEN_AUTO) {
-        saltlen = RSA_size(ctx->rsa) - EVP_MD_get_size(ctx->md) - 2;
+        int mdsize, rsasize;
+
+        if ((mdsize = EVP_MD_get_size(ctx->md)) <= 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+            return -1;
+        }
+        if ((rsasize = RSA_size(ctx->rsa)) <= 0 || rsasize < mdsize + 2) {
```
Comment (tom-cosgrove-arm) [inline providers/implementations/signature/rsa_sig.c:237]: Note that this isn't totally safe: with extreme values for `rsasize` and `mdsize` this test isn't sufficient. For example, a quick local test

```
$ cat overflow.c
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char *argv[])
{
    int rsasize = atoi(argv[1]);
    int mdsize = atoi(argv[2]);

    if (rsasize < mdsize + 2) {
        printf("rsasize < mdsize + 2\n");
        exit(1);
    }

    int saltlen = rsasize - mdsize - 2;
    printf("saltlen = rsasize - mdsize - 2 = %d\n", saltlen);
}

$ make overflow
cc     overflow.c   -o overflow

$ ./overflow 2147483647 2147483647
saltlen = rsasize - mdsize - 2 = -2
```

A simple solution might be to limit the maximum values of `mdsize` and `rsasize` to something well within the limits of `int` but well outside the bounds of reasonableness for what they are - e.g. 1048576.

**Review by tom-cosgrove-arm — APPROVED** [review]
Okay as it stands, but one comment

File: providers/implementations/signature/rsa_sig.c (hunk 1 above)
Comment (paulidale) [reply providers/implementations/signature/rsa_sig.c:237]: `rsasize < 2 || rsasize - 2 < mdsize` works I think.
An alternative would be casting to unsigned or size_t with appropriate checks for negatives.

Comment (t8m) [reply providers/implementations/signature/rsa_sig.c:237]: mdsize+2 cannot really overflow unless the provider is malicious.

**Comment by t8m** [comment]
@tom-cosgrove-arm @beldmit please reconfirm

File: providers/implementations/signature/rsa_sig.c (hunk 1 above)
Comment (t8m) [reply providers/implementations/signature/rsa_sig.c]: fixup pushed

**Review by tom-cosgrove-arm — APPROVED** [review]
Reconfirming - thanks

**Comment by beldmit** [comment]
Reconfirming - thanks

**Comment by t8m** [comment]
Merged to the master, 3.3, 3.2 and 3.1 branches. Thank you for the reviews.

**Pull request closed without merging**